src/hv_aic_events.c

```diff
@@ -0,0 +1,166 @@
+/* SPDX-License-Identifier: MIT */
+
+#include "adt.h"
+#include "hv.h"
+#include "uartproxy.h"
+
+#define AIC_EVENT      0x2004
+#define AIC_EVENT_TYPE GENMASK(31, 16)
+#define AIC_EVENT_NUM  GENMASK(15, 0)
+
+#define AIC_EVENT_TYPE_HW   1
+#define AIC_EVENT_TYPE_IPI  4
+#define AIC_EVENT_IPI_OTHER 1
+#define AIC_EVENT_IPI_SELF  2
+
+#define AIC_MAX_HW_NUM 1024
+
+static u64 aic_base;
+
+static u8 trace_hw_num[AIC_MAX_HW_NUM / 8];
+
+static void set_hw_num(u16 num)
+{
+    u8 *b = &trace_hw_num[num / 8];
+    *b |= 1 << (num & 0x7);
```

Comment:
This is `BIT(num & 7)`. (or `& 31` when switching to u32), and at that point it's probably better to just do it in one line of code. Same for the ones below.